### PR TITLE
DCOS-37612: Jobs Menu Data Layer - Run Job

### PR DESCRIPTION
**Pull request by bstavroulakis**
This PR has 2 goals:

1) The groundwork to detach the menu from the JobsDetails page.
2) Add the Run Job with the new Data Layer pattern.

**Checklist**
- [x] Did you add a JIRA issue in a commit message or as part of the branch name?
- [ ] Did you add new unit tests?
- [ ] Did you add new integration tests?
- [ ] If this is a regression, did you write a test to catch this in the future?

### CODE REVIEW

**Comment by Fabs** [comment]
@bstavroulakis I am also rebasing it so it is aligned with master.

File: src/js/pages/jobs/JobDetailPage.js [hunk 1]
```diff
@@ -329,10 +331,7 @@ class JobDetailPage extends mixin(StoreMixin, TabsMixin) {
       onItemSelect: this.handleEditButtonClick
     });
 
-    actions.push({
```
Comment (weblancaster) [inline src/js/pages/jobs/JobDetailPage.js]: because the store method isn't used anymore, you should be able to remove `handleRunNowButtonClick` from `JobDetailPage`

**Comment by nLight** [comment]
Ready for review? Jira is in progress

File: plugins/jobs/src/js/JobsMenuClient.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import { makeExecutableSchema } from "graphql-tools/dist/index";
+import { runJob } from "#SRC/js/events/MetronomeClient";
+
+export const typeDefs = `
+  type Job {
+      id: String,
+      jobId: String,
+      status: String,
+      createdAt: String
+  }
```
Comment (Fabs) [inline plugins/jobs/src/js/JobsMenuClient.js]: If we are only using the jobId right now, that would be the only thing I would put on this type.

File: plugins/jobs/src/js/JobsRunNow.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import React from "react";
+/* eslint-enable no-unused-vars */
+
+import { graphqlObservable } from "data-service";
+import gql from "graphql-tag";
+
+import { Subject } from "rxjs/Subject";
+import "rxjs/add/operator/combineLatest";
+import "rxjs/add/operator/do";
+import "rxjs/observable/empty";
```
Comment (Fabs) [inline plugins/jobs/src/js/JobsRunNow.js]: I would only leave here the rxjs imports we use (which probably means removing all the operator/observable imports).

File: plugins/jobs/src/js/JobsRunNow.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+});
+
+export const jobsRunNowAction = id => {
+  return {
+    label: "Run Now",
+    onItemSelect() {
+      jobsRunNow$.next({ jobId: id });
+    }
+  };
+};
```
Comment (Fabs) [inline plugins/jobs/src/js/JobsRunNow.js]: Here we could test that running the callback (onItemSelect()) really triggers the graphql call.

File: plugins/jobs/src/js/JobsMenu.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  const newActions = [];
+
+  newActions.push(edit);
+
+  newActions.push(jobsRunNowAction(props.job.getId()));
+
+  return newActions.concat(rest);
+};
+
+const JobsMenu = componentFromStream(props$ => {
```
Comment (Fabs) [inline plugins/jobs/src/js/JobsMenu.js]: What about snapshot testing this one? We can have the cases where it has the schedule or not.

File: plugins/jobs/src/js/JobsRunNow.js (hunk 4 above)
Comment (bstavroulakis) [reply plugins/jobs/src/js/JobsRunNow.js]: When I run the method from the test, for some reason the .next doesn't make the graphql call. Will need to research this more.

File: plugins/jobs/src/js/JobsMenu.js (hunk 5 above)
Comment (bstavroulakis) [reply plugins/jobs/src/js/JobsMenu.js]: Also, in the test with the snapshot the componentFromStream doesn't initialize correctly. Will need some investigation.

File: plugins/jobs/src/js/JobsMenu.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/* eslint-disable no-unused-vars */
+import React from "react";
+/* eslint-enable no-unused-vars */
+
+import { componentFromStream } from "data-service";
+
+import "rxjs/add/operator/combineLatest";
```
Comment (weblancaster) [inline plugins/jobs/src/js/JobsMenu.js]: this line and the 2 below below aren't in use so they can be removed

File: plugins/jobs/src/js/JobsMenu.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+  newActions.push(edit);
+
+  newActions.push(jobsRunNowAction(props.job.getId()));
+
+  return newActions.concat(rest);
+};
+
+const JobsMenu = componentFromStream(props$ => {
+  return props$.combineLatest(jobsRunNow$.startWith({}), props => {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobsMenu.js]: Why do we need `jobRunNow$`? I can't see that we are utilizing any of its observables output

File: plugins/jobs/src/js/__tests__/JobsRunNow-test.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+describe("Jobs Run Now Action", function() {
+  it("runNowJobGraphql call if made with onItemSelect", function(done) {
+    const runNowJobGraphql = jest.fn().mockReturnValue(
+      new Observable(() => {
+        expect(runNowJobGraphql).toBeCalled();
+        done();
+      })
+    );
+
+    const jobsRunNowAction = jobsRunNow.jobsRunNowAction(
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/__tests__/JobsRunNow-test.js]: Do I get that right that we have no test which verifies that the GraphQL schema goes together with what we expect the client to deliver?

File: plugins/jobs/src/js/JobsRunNow.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+export const jobsRunNow$ = runNowEvent$.switchMap(jobId => {
+  return runNowJobGraphql(jobId);
+});
+
+export const jobsRunNowAction = (id, externalRunNowJobGraphql) => {
+  return {
+    label: "Run Now",
+    onItemSelect() {
+      if (externalRunNowJobGraphql) {
+        runNowJobGraphql = externalRunNowJobGraphql;
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobsRunNow.js]: This seems like a super hacky solution for testing to me. I would rather like to have the respective functions hoisted to achieve something like this.

File: plugins/jobs/src/js/JobsRunNow.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+import { graphqlObservable } from "data-service";
+import gql from "graphql-tag";
+
+import { Subject } from "rxjs/Subject";
+import { defaultSchema } from "./JobsMenuClient";
+
+const runNowJobMutation = gql`
+  mutation {
+    runJob(jobId: $jobId) {
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobsRunNow.js]: How does GraphQL know what $jobId is? You only pass a variable to the graphqlObservable, right?

File: plugins/jobs/src/js/__tests__/JobsRunNow-test.js [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+      "test",
+      runNowJobGraphql
+    );
+
+    jobsRunNow.jobsRunNow$.forEach(() => {});
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/__tests__/JobsRunNow-test.js]: What's the purpose of this?

File: webpack/webpack.dev.js [hunk 12]
```diff
@@ -13,7 +13,7 @@ module.exports = merge(common, {
     index: "./src/js/index.js",
     vendor: Object.keys(dependencies)
   },
-  devtool: "cheap-module-eval-source-map",
+  devtool: "source-map",
```
Comment (DanielMSchmidt) [inline webpack/webpack.dev.js]: This seems unrelated, could you pull this into a seperate PR and open a ticket for why this is needed?

File: plugins/jobs/src/js/JobsMenuClient.js [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+import { makeExecutableSchema } from "graphql-tools/dist/index";
+import { runJob } from "#SRC/js/events/MetronomeClient";
+
+export const typeDefs = `
+  type Job {
+      jobId: String
+  }
+
+  type Mutation {
+    runJob(jobId: String!): Job!
```
Comment (DanielMSchmidt) [inline plugins/jobs/src/js/JobsMenuClient.js]: The return type is `JobRun` and not Job. For a job the id would be just id. As I am currently working on typing this stuff out and transforming it into something with the shape of a struct I think a renaming here would be suficcient. Otherwise we would do double work if you would adapt the schema here and did the data transformation

Comment (bstavroulakis) [reply plugins/jobs/src/js/JobsMenuClient.js]: I changed the type to JobRun and the ids to id.

File: webpack/webpack.dev.js (hunk 12 above)
Comment (bstavroulakis) [reply webpack/webpack.dev.js]: Oops, I do this to make debug work in VSCode. Had an auto-ignore script that probably changed now that the webpack config changed and this slipped. No need for us to change this, cheap eval is faster for dev.

File: plugins/jobs/src/js/__tests__/JobsRunNow-test.js (hunk 11 above)
Comment (bstavroulakis) [reply plugins/jobs/src/js/__tests__/JobsRunNow-test.js]: Basically you have to listen to the observable for it to actually start as it is lazy. Changed this to subscribe.

File: plugins/jobs/src/js/JobsRunNow.js (hunk 10 above)
Comment (bstavroulakis) [reply plugins/jobs/src/js/JobsRunNow.js]: Changed to id: $id. I'm not 100% how does this work, but if I put id: String or something else it doesn't map correctly with the defaultSchema. Probably @Fabs can help us here.

File: plugins/jobs/src/js/JobsRunNow.js (hunk 9 above)
Comment (bstavroulakis) [reply plugins/jobs/src/js/JobsRunNow.js]: Yes, was frustrated and wanted to make this work. Definitely not optimal, I changed this.

File: plugins/jobs/src/js/__tests__/JobsRunNow-test.js (hunk 8 above)
Comment (bstavroulakis) [reply plugins/jobs/src/js/__tests__/JobsRunNow-test.js]: Yeah, no test for that. Would we want a test for this?

File: npm-shrinkwrap.json [hunk 14]
```diff
@@ -1150,12 +1150,6 @@
       "integrity": "sha512-e74sM9W/4qqWB6D4TWV9FQk0WoHtX1X4FJpbjxucMSVJHtFjbQOH3H6yp+xno4br0AKG0wz/kPtaN599GUOvAg==",
       "dev": true
     },
-    "@types/jest": {
```
[the remark this answers is not in the document]
Comment (bstavroulakis) [reply npm-shrinkwrap.json]: Hmm, I just did ```rm -rf node_modules``` and ran ```npm install``` again. This is odd.

[the remark this answers is not in the document]
Comment (bstavroulakis) [reply npm-shrinkwrap.json]: I will revert to the old one.

File: plugins/jobs/src/js/JobsMenu.js (hunk 7 above)
Comment (bstavroulakis) [reply plugins/jobs/src/js/JobsMenu.js]: When I remove this and just do ```props$.subscribe(props => {})``` I get ```cannot use props$ undefined```. I guess we do this to extend with an observable that always exist. Maybe @Fabs can clear this out a bit.

File: plugins/jobs/src/js/JobsRunNow.js [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  }
+`;
+
+const runNowJobGraphql = id => {
+  return graphqlObservable(runNowJobMutation, defaultSchema, id);
+};
+
+export const runNowEvent$ = new Subject();
+export const jobsRunNow$ = runNowEvent$.switchMap(id => {
```
Comment (bstavroulakis) [inline plugins/jobs/src/js/JobsRunNow.js]: @DanielMSchmidt Over here we use switchmap with the runNowEvent Subject to actually listen to this observable but also have the ability to trigger it as well.

File: plugins/jobs/src/js/JobsMenu.js (hunk 7 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobsMenu.js]: `combineLatest` will not emit unless both the streams it is combining emits at least once. So if you combinLatest the props with jobsRunNow, the menu will never appear. You need to give jobsRunNow an initial value.

File: plugins/jobs/src/js/JobsMenuClient.js [hunk 16]
```diff
@@ -0,0 +1,32 @@
+import "rxjs/add/operator/combineLatest";
+import "rxjs/add/operator/map";
+
+import { makeExecutableSchema } from "graphql-tools/dist/index";
+import { runJob } from "#SRC/js/events/MetronomeClient";
+
+export const typeDefs = `
```
Comment (Fabs) [inline plugins/jobs/src/js/JobsMenuClient.js]: Now that Daniel has this schema:
https://github.com/dcos/dcos-ui/pull/3006

I suggest we make this queries against that schema, and move our resolvers and mutations to that file (I believe it is a good idea to rebase from that branch).

Also the naming of this file is a little of. The files with extension client are usually network to data layer adapters. The files with graphql resolvers and type definitions are supposed to be Models. And we are aiming to have an unified model for jobs.

File: plugins/jobs/src/js/JobsMenu.js (hunk 7 above)
Comment (DanielMSchmidt) [reply plugins/jobs/src/js/JobsMenu.js]: I get that, but `jobRunNow$` is called with `.next({id})` and this id property is not used here as far as I can see, so why do we need the value of the stream?

File: plugins/jobs/src/js/JobsMenuClient.js (hunk 16 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobsMenuClient.js]: Moving that I also think we should test the resolvers for
- returning a stream
- the stream has the transformations I want (which in this case might mean just that the runJob is being called with the id of the args I passed.

File: plugins/jobs/src/js/JobsMenu.js (hunk 7 above)
Comment (Fabs) [reply plugins/jobs/src/js/JobsMenu.js]: You need to pass the id to the Subject, to trigger the runNow of the particular id. 

jobRunNow$ is a stream of all the results of the executions of the graphqlMutation to run a job. To issue the command you have to onNext to the Subject related to that, so the subject will switchMap into the request.

**Comment by bstavroulakis** [comment]
Thanks @DanielMSchmidt for the review, fixed conflicts.

File: src/js/pages/jobs/JobDetailPage.js [hunk 17]
```diff
@@ -245,10 +242,12 @@ class JobDetailPage extends mixin(TabsMixin) {
 
     return (
       <Page>
-        <Page.Header
+        <JobsMenu
```
Comment (nLight) [inline src/js/pages/jobs/JobDetailPage.js]: Please consider keeping `Page.Header` looks like the way @orlandohohmeier suggested in the thread https://mesosphere.slack.com/archives/C03UXUVH7/p1529066379000323 could work. With that we won't need to modify the TemplateUtil

**Comment by DanielMSchmidt** [comment]
@bstavroulakis This can run against master now <3 sorry for all the conflicts 😞

**Comment by Fabs** [comment]
Closing that in favor of #3023, will incorporate the comments here, there.
